src/dataframe/dataframe.jl

```diff
@@ -596,17 +596,25 @@ Base.setindex!(df::DataFrame, x::Void, col_ind::Int) = delete!(df, col_ind)
 
 Base.empty!(df::DataFrame) = (empty!(df.columns); empty!(index(df)); df)
 
-function Base.insert!(df::DataFrame, col_ind::Int, item::AbstractVector, name::Symbol)
+function Base.insert!(df::DataFrame, col_ind::Int, item::AbstractVector,
+                      name::Symbol; allow_duplicates=true)
+    Base.depwarn(string("Currently insert! allows duplicate column names.",
```

Comment:
The idea with deprecations is that they should only be printed when people use the old feature, not all the time. Here, it should only be printed when `allow_duplicates=true` and that a duplicate column name is encountered.

Actually, it's not clear we need to add the `allow_duplicates` argument: people should stop creating columns with duplicate names anyway so that their code continues to work in the next major release. And passing `allow_duplicates=false` isn't really essential, since it will just turn a deprecation warning into an error.

Reply:
> people should stop creating columns with duplicate names anyway

Agreed, but then I would remove `allow_duplicates` argument from `names!` also. If this is OK I will change this PR to only throw `depwarn` in `names!` and `insert!` now when duplicates are encountered (I understand this is the proper way to implement it) - this PR would be merged now. And then have a second PR removing `allow_duplicates` argument from `names!` and making `insert!` disallow duplicates that would be merged after the next DataFrames release.

Is this the way you see we should go?

Reply:
Yes. I guess there is no *real* use case. The only situation I have encountered is that when source data has duplicate columns - but it was always a bug in source data, and actually catching it was what was the proper way to go.

I guess we should treat `DataFrame` as `Dict` if indexed by column name. This is an important invariant. So duplicates should not be allowed.

I will adjust this PR accordingly and make this next PR on top of it to be merged later.